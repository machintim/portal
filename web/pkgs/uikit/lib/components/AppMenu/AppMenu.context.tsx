import React, {
  createContext,
  useContext,
  useState,
  useRef,
  useEffect,
} from 'react';
import useMediaQuery from '../../../src/hooks/useMediaQuery';
import { useTheme } from '../../theme';

interface IContext {
  isOpen: boolean;
  toggleMenu: () => void;
  triggerRef: React.RefObject<HTMLButtonElement | null>;
  contentRef: React.RefObject<HTMLUListElement | null>;
}

export const AppMenuContext = createContext({} as IContext);

export const useAppMenu = () => useContext(AppMenuContext);

const AppMenuProvider = ({
  defaultOpen = false,
  children,
}: {
  defaultOpen?: boolean;
  children: React.ReactNode;
}) => {
  const [isOpen, setIsOpen] = useState(defaultOpen);
  const [theme] = useTheme();
  const isDesktop = useMediaQuery(
    `(min-width: ${theme.breakpoints.desktop}px)`
  );

  const triggerRef = useRef<HTMLButtonElement>(null);
  const contentRef = useRef<HTMLUListElement>(null);

  useEffect(() => {
    const originalStyle = document.body.style.overflow;
    if (isOpen && !isDesktop) {
      // Prevent background scrolling when menu is open and on mobile
      document.body.style.overflow = 'hidden';
    } else {
      document.body.style.overflow = originalStyle;
    }
    return () => {
      document.body.style.overflow = originalStyle;
    };
  }, [isOpen, isDesktop]);

  useEffect(() => {
    if (isOpen && contentRef.current) {
      contentRef.current.focus();

      if (contentRef.current && document.activeElement === contentRef.current) {
        console.log('Menu successfully focused');
      } else {
        console.log(
          'Failed to focus menu - current focus is on:',
          document.activeElement
        );
      }
    }
  }, [isOpen]);

  useEffect(() => {
    const closeMenuOutsideClick = (event: MouseEvent) => {
      if (
        triggerRef.current &&
        !triggerRef.current.contains(event.target as Node) &&
        contentRef.current &&
        !contentRef.current.contains(event.target as Node)
      ) {
        setIsOpen(false);
      }
    };

    const handleKeyDown = (event: KeyboardEvent) => {
      if (event.key === 'Escape') {
        setIsOpen(false);
      }

      if (!isOpen || !contentRef.current) return;

      const focusableElements = Array.from(
        contentRef.current.querySelectorAll('[role="menuitem"]')
      ).filter((el) => {
        const element = el as HTMLElement;
        return (
          element.getAttribute('aria-disabled') !== 'true' &&
          element.tabIndex >= 0 &&
          element.offsetParent !== null
        );
      }) as HTMLElement[];

      if (focusableElements.length === 0) return;

      const firstElement = focusableElements[0];
      const lastElement = focusableElements[focusableElements.length - 1];

      const currentIndex = focusableElements.indexOf(
        document.activeElement as HTMLElement
      );

      if (event.key === 'Tab') {
        // Loop focus on Tab/Shift+Tab
        if (event.shiftKey && document.activeElement === firstElement) {
          lastElement.focus();
          event.preventDefault();
        } else if (!event.shiftKey && document.activeElement === lastElement) {
          firstElement.focus();
          event.preventDefault();
        }
      } else if (event.key === 'ArrowDown') {
        // Move focus to next item
        if (currentIndex !== -1) {
          const nextIndex = (currentIndex + 1) % focusableElements.length;
          focusableElements[nextIndex].focus();
          event.preventDefault();
        } else {
          firstElement.focus();
          event.preventDefault();
        }
      } else if (event.key === 'ArrowUp') {
        // Move focus to previous item
        if (currentIndex !== -1) {
          const prevIndex =
            (currentIndex - 1 + focusableElements.length) %
            focusableElements.length;
          focusableElements[prevIndex].focus();
          event.preventDefault();
        } else {
          lastElement.focus();
          event.preventDefault();
        }
      }
    };

    document.addEventListener('mousedown', closeMenuOutsideClick);
    document.addEventListener('keydown', handleKeyDown);

    return () => {
      document.removeEventListener('mousedown', closeMenuOutsideClick);
      document.removeEventListener('keydown', handleKeyDown);
    };
  }, [isOpen]);

  const toggleMenu = () => setIsOpen((prevOpen) => !prevOpen);

  const value = {
    isOpen,
    toggleMenu,
    triggerRef,
    contentRef,
  };

  return (
    <AppMenuContext.Provider value={value}>{children}</AppMenuContext.Provider>
  );
};

export default AppMenuProvider;
