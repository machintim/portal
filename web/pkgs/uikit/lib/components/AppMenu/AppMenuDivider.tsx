import React, { HtmlHTMLAttributes } from 'react';
import { css, cx } from '@emotion/css';
import useTheme from '../../theme/useTheme';

export const NAME = 'ucl-app-menu__divider';

export interface AppMenuDividerProps
  extends Omit<
    HtmlHTMLAttributes<HTMLHRElement>,
    'children'
  > {
  testId?: string;
}

const AppMenuDivider: React.FC<AppMenuDividerProps> = ({
  testId = NAME,
  className,
  ...props
}) => {
  const [theme] = useTheme();

  const baseStyle = css`
    margin: 0;
    height: 1px;
    border: none;
    background-color: ${theme.color.neutral.grey20};
  `;

  const style = cx(NAME, baseStyle, className);

  return (
    <hr
      className={style}
      data-testid={testId}
      {...props}
    />
  );
};

export default AppMenuDivider;
