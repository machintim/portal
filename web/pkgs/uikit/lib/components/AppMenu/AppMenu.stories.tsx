import type { Meta, StoryObj } from '@storybook/react';
import { Link as InternalLink } from 'wouter';
import AppMenu from './AppMenu';
import Icon from '../Icon/Icon';
import Link from '../Link/Link';

const meta = {
  title: 'Components/Deprecated/AppMenu',
  component: AppMenu,
} satisfies Meta<typeof AppMenu>;

export default meta;

type Story = StoryObj<typeof meta>;

export const Simple: Story = {
  render: () => {
    return (
      <AppMenu>
        <AppMenu.MenuItem>Menu item 01</AppMenu.MenuItem>
        <AppMenu.MenuItem>Menu item 02</AppMenu.MenuItem>
        <AppMenu.MenuItem>Menu item 03</AppMenu.MenuItem>
      </AppMenu>
    );
  },
};

export const WithDividers: Story = {
  name: 'With dividers',
  render: () => {
    return (
      <AppMenu>
        <AppMenu.MenuItem>Menu item 01</AppMenu.MenuItem>
        <AppMenu.Divider />
        <AppMenu.MenuItem>Menu item 02</AppMenu.MenuItem>
        <AppMenu.Divider />
        <AppMenu.MenuItem>Menu item 03</AppMenu.MenuItem>
      </AppMenu>
    );
  },
};

export const WithIcons: Story = {
  name: 'With icons',
  render: () => {
    return (
      <AppMenu>
        <AppMenu.MenuItem icon={<Icon.Calendar size={20} />}>
          Calendar
        </AppMenu.MenuItem>
        <AppMenu.MenuItem icon={<Icon.Edit size={20} />}>Edit</AppMenu.MenuItem>
        <AppMenu.MenuItem icon={<Icon.Grid size={20} />}>Apps</AppMenu.MenuItem>
        <AppMenu.MenuItem icon={<Icon.Settings size={20} />}>
          Settings
        </AppMenu.MenuItem>
      </AppMenu>
    );
  },
};

export const WithTrigger: Story = {
  name: 'With trigger',
  render: () => {
    return (
      <AppMenu>
        <AppMenu.Trigger>
          <Icon.Menu />
        </AppMenu.Trigger>
        <AppMenu.Content>
          <AppMenu.MenuItem>Menu item 01</AppMenu.MenuItem>
          <AppMenu.MenuItem>Menu item 02</AppMenu.MenuItem>
          <AppMenu.MenuItem>Menu item 03</AppMenu.MenuItem>
        </AppMenu.Content>
      </AppMenu>
    );
  },
};

export const Complex: Story = {
  render: () => (
    <AppMenu>
      <AppMenu.MenuItem icon={<Icon.Calendar size={20} />}>
        <InternalLink
          to='/timetable/personal'
          asChild
        >
          <Link noVisited>Personal Timetable</Link>
        </InternalLink>
      </AppMenu.MenuItem>
      <AppMenu.MenuItem icon={<Icon.Calendar size={20} />}>
        <InternalLink
          to='/timetable/degree'
          asChild
        >
          <Link noVisited>Degree Timetable</Link>
        </InternalLink>
      </AppMenu.MenuItem>
      <AppMenu.MenuItem icon={<Icon.Calendar size={20} />}>
        <InternalLink
          to='/timetable/custom'
          asChild
        >
          <Link noVisited>Custom Timetable</Link>
        </InternalLink>
      </AppMenu.MenuItem>
      <AppMenu.Divider />
      <AppMenu.MenuItem icon={<Icon.Settings size={20} />}>
        Settings
      </AppMenu.MenuItem>
      <AppMenu.MenuItem icon={<Icon.HelpCircle size={20} />}>
        Help
      </AppMenu.MenuItem>
      <AppMenu.Divider />
      <AppMenu.MenuItem icon={<Icon.ArrowLeft size={20} />}>
        <InternalLink
          to='/timetable/home'
          asChild
        >
          <Link noVisited>Exit beta</Link>
        </InternalLink>
      </AppMenu.MenuItem>
    </AppMenu>
  ),
};

export const ComplexWithTrigger: Story = {
  name: 'Complex with trigger',
  render: () => (
    <AppMenu>
      <AppMenu.Trigger>
        <Icon.Menu />
      </AppMenu.Trigger>
      <AppMenu.Content>
        <AppMenu.MenuItem icon={<Icon.Calendar size={20} />}>
          <InternalLink
            to='/timetable/personal'
            asChild
          >
            <Link noVisited>Personal Timetable</Link>
          </InternalLink>
        </AppMenu.MenuItem>
        <AppMenu.MenuItem icon={<Icon.Calendar size={20} />}>
          <InternalLink
            to='/timetable/degree'
            asChild
          >
            <Link noVisited>Degree Timetable</Link>
          </InternalLink>
        </AppMenu.MenuItem>
        <AppMenu.MenuItem icon={<Icon.Calendar size={20} />}>
          <InternalLink
            to='/timetable/custom'
            asChild
          >
            <Link noVisited>Custom Timetable</Link>
          </InternalLink>
        </AppMenu.MenuItem>
        <AppMenu.Divider />
        <AppMenu.MenuItem icon={<Icon.Settings size={20} />}>
          Settings
        </AppMenu.MenuItem>
        <AppMenu.MenuItem icon={<Icon.HelpCircle size={20} />}>
          Help
        </AppMenu.MenuItem>
        <AppMenu.Divider />
        <AppMenu.MenuItem icon={<Icon.ArrowLeft size={20} />}>
          <InternalLink
            to='/timetable/home'
            asChild
          >
            <Link noVisited>Exit beta</Link>
          </InternalLink>
        </AppMenu.MenuItem>
      </AppMenu.Content>
    </AppMenu>
  ),
};
