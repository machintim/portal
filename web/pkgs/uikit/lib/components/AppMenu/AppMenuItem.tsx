import React, { LiHTMLAttributes } from 'react';
import { css, cx } from '@emotion/css';
import Link from '../Link';
import { Link as InternalLink, useLocation } from 'wouter';
import { useAppMenu } from './AppMenu.context';
import useTheme from '../../theme/useTheme';

export const NAME = 'ucl-app-menu__item';

export interface AppMenuItemProps extends LiHTMLAttributes<HTMLLIElement> {
  testId?: string;
  icon?: React.ReactNode;
  internalLink?: string;
  externalLink?: string;
}

const AppMenuItem: React.FC<AppMenuItemProps> = ({
  testId = NAME,
  icon,
  children,
  className,
  internalLink,
  externalLink,
  onClick,
  ...props
}) => {
  const isClickable = !!(onClick || internalLink || externalLink);

  const [theme] = useTheme();
  const { toggleMenu } = useAppMenu();
  const [location] = useLocation();
  const isCurrent = internalLink ? location === internalLink : false;

  const handleClick = (e: React.MouseEvent<Element, MouseEvent>) => {
    if (isCurrent) {
      e.preventDefault();
      toggleMenu(); // Close the menu
    }
    if (onClick) {
      onClick(e as unknown as React.MouseEvent<HTMLLIElement, MouseEvent>);
    }
  };

  const baseStyle = css`
    height: 40px;
    padding: 0 16px;
    display: flex;
    align-items: center;
    font-family: ${theme.font.family.primary};
    font-size: ${theme.font.size.f16};
    color: ${theme.color.text.secondary};
  `;

  const clickableMenuItemStyle = css`
    color: ${theme.color.text.primary};
    &:hover {
      background-color: ${theme.color.interaction.blue5};
    }
    &:focus-visible {
      outline: none;
      box-shadow: ${theme.boxShadow.focus};
      z-index: 1;
    }
  `;

  const menuLinkStyle = css`
    text-decoration: none;
    color: ${theme.color.text.primary};
    display: flex;
    align-items: center;
    width: 100%;
    height: 100%;
    &:hover {
      background-color: ${theme.color.interaction.blue5};
    }
    &:visited {
      color: ${theme.color.text.primary};
    }
    &:focus-visible {
      outline: none;
      box-shadow: ${theme.boxShadow.focus};
      z-index: 1;
    }
  `;

  const iconContainerStyle = css`
    display: inline-flex;
    align-items: center;
    margin-right: 8px;
  `;

  const style = cx(
    NAME,
    baseStyle,
    isClickable && clickableMenuItemStyle,
    className
  );

  const content = (
    <>
      {React.isValidElement(icon) && (
        <span className={iconContainerStyle}>{icon}</span>
      )}
      {children}
    </>
  );

  if (internalLink) {
    // if a link is provided, return a list item with a link
    return (
      <InternalLink
        to={internalLink}
        asChild
        onClick={handleClick}
      >
        <Link
          noVisited
          className={menuLinkStyle}
          role='menuitem'
          tabIndex={0}
          aria-label={
            isCurrent ? `${children} - Currently selected` : undefined
          }
        >
          <li
            className={style}
            data-testid={testId}
            {...props}
          >
            {content}
          </li>
        </Link>
      </InternalLink>
    );
  }

  if (externalLink) {
    // if an external link is provided, return a list item with an external link
    return (
      <Link
        href={externalLink}
        noVisited
        className={menuLinkStyle}
        role='menuitem'
        tabIndex={0}
      >
        <li
          className={style}
          data-testid={testId}
          {...props}
        >
          {content}
        </li>
      </Link>
    );
  }

  return (
    // if no link is provided, return a simple list item
    <li
      tabIndex={isClickable ? 0 : -1}
      role='menuitem'
      aria-disabled={!isClickable}
      className={style}
      data-testid={testId}
      onClick={onClick}
      {...props}
    >
      {content}
    </li>
  );
};

export default AppMenuItem;
