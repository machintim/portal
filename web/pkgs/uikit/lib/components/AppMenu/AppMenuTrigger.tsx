import React from 'react';
import { css, cx } from '@emotion/css';
import { useTheme } from '../../theme';
import IconButton, { IconButtonProps } from '../IconButton';
import { useAppMenu } from './AppMenu.context';

export const NAME = 'ucl-app-menu__trigger';

export interface AppMenuTriggerProps
  extends IconButtonProps {}

const AppMenuTrigger: React.FC<AppMenuTriggerProps> = ({
  testId = NAME,
  className,
  children,
  ...props
}) => {
  const [theme] = useTheme();
  const { isOpen, toggleMenu, triggerRef } = useAppMenu();

  const baseStyle = css`
    &:focus-visible {
      outline: none;
      box-shadow: ${theme.boxShadow.focus};
      z-index: 1;
    }
  `;

  const style = cx(NAME, baseStyle, className);

  return (
    <IconButton
      id='app-menu-button'
      ref={triggerRef}
      className={style}
      onClick={toggleMenu}
      aria-label='Menu button'
      aria-controls='app-menu'
      aria-haspopup='true'
      aria-expanded={isOpen ? 'true' : 'false'}
      data-testid={testId}
      tabIndex={0}
      {...props}
    >
      {children}
    </IconButton>
  );
};

export default AppMenuTrigger;
