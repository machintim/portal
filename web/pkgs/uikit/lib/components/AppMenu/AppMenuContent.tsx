import React, { LiHTMLAttributes, useEffect } from 'react';
import { css, cx } from '@emotion/css';
import useTheme from '../../theme/useTheme';
import { useAppMenu } from './AppMenu.context';
import AppMenuItem from './AppMenuItem';
import Icon from '../Icon';
import { createPortal } from 'react-dom';
import useMediaQuery from '../../../src/hooks/useMediaQuery';

export const NAME = 'ucl-app-menu__content';

export interface AppMenuContentProps
  extends LiHTMLAttributes<HTMLUListElement> {
  testId?: string;
}

const AppMenuContent: React.FC<AppMenuContentProps> = ({
  testId = NAME,
  children,
  className,
  ...props
}) => {
  const { isOpen, contentRef, toggleMenu } = useAppMenu();
  const [theme] = useTheme();

  const isDesktop = useMediaQuery(
    `(min-width: ${theme.breakpoints.desktop}px)`
  );

  const handleBackButtonKeyDown = (e: React.KeyboardEvent) => {
    if (e.key === 'Enter') {
      toggleMenu();
    }
  };

  const baseStyle = css`
    z-index: 9999;
    position: absolute;
    display: none;
    top: 16px;
    min-width: 140px;
    white-space: nowrap;
    cursor: default;
    background-color: white;
    box-shadow: ${theme.boxShadow.y1};
    border: ${theme.border.b1} solid ${theme.color.neutral.grey20};
    padding-left: 0;
    padding-right: 0;
    color: ${theme.color.text.primary};
    list-style: none;

    @media (max-width: ${theme.breakpoints.desktop}px) {
      position: fixed;
      top: 0;
      left: 0;
      width: 100vw;
      height: 100vh;
      border: none;
      box-shadow: none;
      padding: 0;
      margin: 0;
      overflow-y: auto;
      a {
        height: auto;
      }
    }
  `;

  const expandedStyle = css`
    display: block;
  `;

  const backMenuItemStyle = css`
    cursor: pointer;

    @media (min-width: ${theme.breakpoints.desktop}px) {
      display: none;
    }
  `;

  useEffect(() => {
    const mainDiv = document.getElementById('root');
    if (mainDiv) {
      if (!isDesktop && isOpen) {
        mainDiv.style.display = 'none';
      } else {
        mainDiv.style.display = '';
      }
    }
  }, [isOpen, isDesktop]);

  const style = cx(NAME, baseStyle, isOpen && expandedStyle, className);

  const menuContent = (
    <ul
      id='app-menu'
      role='menu'
      aria-labelledby='app-menu-button'
      ref={contentRef}
      className={style}
      data-testid={testId}
      tabIndex={-1}
      {...props}
    >
      <AppMenuItem
        onClick={toggleMenu}
        onKeyDown={handleBackButtonKeyDown}
        icon={<Icon.ChevronLeft />}
        className={backMenuItemStyle}
        aria-label='Go Back'
      >
        Back
      </AppMenuItem>
      {children}
    </ul>
  );

  return !isDesktop ? createPortal(menuContent, document.body) : menuContent;
};

export default AppMenuContent;
