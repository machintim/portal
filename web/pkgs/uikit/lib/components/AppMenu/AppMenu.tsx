import { HTMLAttributes } from 'react';
import { css, cx } from '@emotion/css';
import { useTheme } from '../../theme';
import AppMenuProvider from './AppMenu.context';
import AppMenuContent from './AppMenuContent';
import AppMenuItem from './AppMenuItem';
import AppMenuDivider from './AppMenuDivider';
import AppMenuTrigger from './AppMenuTrigger';

export const NAME = 'ucl-app-menu';

export interface AppMenuProps
  extends HTMLAttributes<HTMLDivElement> {
  testId?: string;
  defaultOpen?: boolean;
}

const AppMenu = ({
  testId = NAME,
  defaultOpen = false,
  children,
  className,
  ...props
}: AppMenuProps) => {
  const [theme] = useTheme();

  const baseStyle = css`
    position: relative;
    font-family: ${theme.font.family.primary};
  `;

  const style = cx(NAME, baseStyle, className);

  return (
    <div
      className={style}
      data-testid={testId}
      {...props}
    >
      <AppMenuProvider defaultOpen={defaultOpen}>
        {children}
      </AppMenuProvider>
    </div>
  );
};

export interface IAppMenuSubComponents {
  Trigger: typeof AppMenuTrigger;
  Content: typeof AppMenuContent;
  MenuItem: typeof AppMenuItem;
  Divider: typeof AppMenuDivider;
}

const AppMenuWithSubComponents = AppMenu as typeof AppMenu &
  IAppMenuSubComponents;

AppMenuWithSubComponents.Trigger = AppMenuTrigger;
AppMenuWithSubComponents.Content = AppMenuContent;
AppMenuWithSubComponents.MenuItem = AppMenuItem;
AppMenuWithSubComponents.Divider = AppMenuDivider;

export default AppMenuWithSubComponents;
