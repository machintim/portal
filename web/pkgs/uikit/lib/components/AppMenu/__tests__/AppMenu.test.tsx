import { describe, expect, test, beforeAll, vi } from 'vitest';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import '@testing-library/jest-dom';
import AppMenu from '../AppMenu';
import { ThemeContextProvider } from '../../../theme/useTheme';
import Icon from '../../Icon';

beforeAll(() => {
  globalThis.matchMedia = vi.fn().mockImplementation((query) => ({
    matches: query === '(min-width: 1056px)',
    media: query,
    addListener: vi.fn(),
    removeListener: vi.fn(),
    addEventListener: vi.fn(),
    removeEventListener: vi.fn(),
  }));
});

describe('AppMenu', () => {
  //snapshot
  test('snapshot: no props', () => {
    const renderResult = render(
      <ThemeContextProvider>
        <AppMenu />
      </ThemeContextProvider>
    );
    expect(renderResult.container.firstChild).toMatchSnapshot();
  });

  test('snapshot: testId prop', () => {
    const renderResult = render(
      <ThemeContextProvider>
        <AppMenu testId='test1' />
      </ThemeContextProvider>
    );
    expect(renderResult.container.firstChild).toMatchSnapshot();
  });

  test('snapshot: defaultOpen prop', () => {
    const renderResult = render(
      <ThemeContextProvider>
        <AppMenu defaultOpen>
          <AppMenu.Trigger>
            <Icon.Menu />
          </AppMenu.Trigger>
          <AppMenu.Content>
            <AppMenu.MenuItem>Menu Item 1</AppMenu.MenuItem>
            <AppMenu.Divider />
            <AppMenu.MenuItem>Menu Item 2</AppMenu.MenuItem>
          </AppMenu.Content>
        </AppMenu>
      </ThemeContextProvider>
    );
    expect(renderResult.container.firstChild).toMatchSnapshot();
  });

  //interaction
  test('test ID: default', () => {
    render(
      <ThemeContextProvider>
        <AppMenu />
      </ThemeContextProvider>
    );
    const appMenu = screen.getByTestId('ucl-app-menu');
    expect(appMenu).toBeDefined();
  });

  test('menu button click reveals menu list', async () => {
    const user = userEvent.setup();

    render(
      <ThemeContextProvider>
        <AppMenu>
          <AppMenu.Trigger testId='menu-trigger'>
            <Icon.Menu />
          </AppMenu.Trigger>
          <AppMenu.Content testId='menu-content'>
            <AppMenu.MenuItem>Menu Item 1</AppMenu.MenuItem>
          </AppMenu.Content>
        </AppMenu>
      </ThemeContextProvider>
    );

    const appMenuButton = screen.getByTestId(
      'menu-trigger'
    ) as HTMLButtonElement;
    const appMenuContent = screen.getByTestId(
      'menu-content'
    ) as HTMLUListElement;

    const computedStylesBeforeClick = window.getComputedStyle(appMenuContent);
    expect(computedStylesBeforeClick.display).toBe('none');

    await user.click(appMenuButton);

    const computedStylesAfterClick = window.getComputedStyle(appMenuContent);
    expect(computedStylesAfterClick.display).toBe('block');
  });
});
