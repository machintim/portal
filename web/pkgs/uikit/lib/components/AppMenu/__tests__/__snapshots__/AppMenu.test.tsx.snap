// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`AppMenu > snapshot: defaultOpen prop 1`] = `
<div
  class="ucl-app-menu css-it216d"
  data-testid="ucl-app-menu"
>
  <button
    aria-controls="app-menu"
    aria-expanded="true"
    aria-haspopup="true"
    aria-label="Menu button"
    class="ucl-icon-button ucl-app-menu__trigger css-mgnm3l"
    data-testid="ucl-app-menu__trigger"
    id="app-menu-button"
    tabindex="0"
  >
    <svg
      class="ucl-icon css-148hpxb"
      data-testid="ucl-icon"
      fill="none"
      height="24"
      stroke="currentColor"
      stroke-linecap="round"
      stroke-linejoin="round"
      stroke-width="2"
      viewBox="0 0 24 24"
      width="24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <line
        x1="3"
        x2="21"
        y1="12"
        y2="12"
      />
      <line
        x1="3"
        x2="21"
        y1="6"
        y2="6"
      />
      <line
        x1="3"
        x2="21"
        y1="18"
        y2="18"
      />
    </svg>
  </button>
  <ul
    aria-labelledby="app-menu-button"
    class="ucl-app-menu__content css-63xuy3"
    data-testid="ucl-app-menu__content"
    id="app-menu"
    role="menu"
    tabindex="-1"
  >
    <li
      aria-disabled="false"
      aria-label="Go Back"
      class="ucl-app-menu__item css-8z8u2q"
      data-testid="ucl-app-menu__item"
      role="menuitem"
      tabindex="0"
    >
      <span
        class="css-1epmcqg"
      >
        <svg
          class="ucl-icon css-148hpxb"
          data-testid="ucl-icon"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <polyline
            points="15 18 9 12 15 6"
          />
        </svg>
      </span>
      Back
    </li>
    <li
      aria-disabled="true"
      class="ucl-app-menu__item css-vobysi"
      data-testid="ucl-app-menu__item"
      role="menuitem"
      tabindex="-1"
    >
      Menu Item 1
    </li>
    <hr
      class="ucl-app-menu__divider css-wwimuq"
      data-testid="ucl-app-menu__divider"
    />
    <li
      aria-disabled="true"
      class="ucl-app-menu__item css-vobysi"
      data-testid="ucl-app-menu__item"
      role="menuitem"
      tabindex="-1"
    >
      Menu Item 2
    </li>
  </ul>
</div>
`;

exports[`AppMenu > snapshot: no props 1`] = `
<div
  class="ucl-app-menu css-it216d"
  data-testid="ucl-app-menu"
/>
`;

exports[`AppMenu > snapshot: testId prop 1`] = `
<div
  class="ucl-app-menu css-it216d"
  data-testid="test1"
/>
`;
